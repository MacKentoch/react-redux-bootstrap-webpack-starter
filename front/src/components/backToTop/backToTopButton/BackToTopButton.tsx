import React, { ReactNode } from 'react';
import cx from 'classnames';
import UpIcon from './UpIcon';
import WithRightMargin from './styled/WithRightMargin';

// #region types
export type BackButtonPosition = 'bottom-left' | 'bottom-right';
type Props = {
  position: BackButtonPosition;
  onClick: (event: React.MouseEvent<HTMLButtonElement>) => any;
  children?: ReactNode;
  motionStyle: any;
};
// #endregion

// #region constants
const defaultBackGroundColor = '#4A4A4A';
const sideOffset = '-10px';
const bottomOffset = '40px';
const defaultWidth = '100px';
const defaultZindex = 10;
const defaultOpacity = 0.5;
const defaultStyle = {
  position: 'fixed',
  right: sideOffset,
  left: '',
  bottom: bottomOffset,
  width: defaultWidth,
  zIndex: defaultZindex,
  opacity: defaultOpacity,
  backgroundColor: defaultBackGroundColor,
};

function setPosition(position = 'bottom-right', refStyle = defaultStyle): any {
  const style = { ...refStyle };

  switch (position) {
    case 'bottom-right':
      style.right = sideOffset;
      style.left = '';
      return style;

    case 'bottom-left':
      style.right = '';
      style.left = sideOffset;
      return style;

    default:
      return refStyle;
  }
}
// #endregion

const BackToTopButton = ({
  onClick,
  position = 'bottom-right',
  children,
  motionStyle,
}: Props) => {
  const buttonStyle = setPosition(position, {
    ...motionStyle,
    ...defaultStyle,
  });

  return (
    <button
      style={buttonStyle}
      className={cx({
        btn: true,
      })}
      onClick={onClick}
    >
      {!children && (
        <WithRightMargin>
          <UpIcon color={'#F1F1F1'} />
        </WithRightMargin>
      )}
      {!!children && children}
    </button>
  );
};

BackToTopButton.displayName = 'BackToTopButton';

export default BackToTopButton;
