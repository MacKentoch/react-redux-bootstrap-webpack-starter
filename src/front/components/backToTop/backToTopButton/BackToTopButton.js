// @flow

// #region imports
import React from 'react';
import cx from 'classnames';
import UpIcon from './UpIcon';
import WithRightMargin from './styled/WithRightMargin';
// #endregion

// #region flow types
type Props = {
  position: 'bottom-left' | 'bottom-right',
  onClick: (event: SyntheticEvent<>) => any,
  children: any,
  motionStyle: any,
  ...any,
};
// #endregion

// #region constants
const defaultBackGroundColor = '#4A4A4A';
const sideOffset = '-10px';
const bottomOffset = '40px';
const defaultWidth = '100px';
const defaultZindex = 10;
const defaultOpacity = 0.5;
const defaultStyle = {
  position: 'fixed',
  right: sideOffset,
  left: '',
  bottom: bottomOffset,
  width: defaultWidth,
  zIndex: defaultZindex,
  opacity: defaultOpacity,
  backgroundColor: defaultBackGroundColor,
};
// #endregion

const BackToTopButton = ({
  onClick,
  position,
  children,
  motionStyle,
}: Props) => {
  const buttonStyle = setPosition(position, {
    ...motionStyle,
    ...defaultStyle,
  });

  return (
    <button
      style={buttonStyle}
      className={cx({
        btn: true,
      })}
      onClick={onClick}
    >
      {!children && (
        <WithRightMargin>
          <UpIcon color={'#F1F1F1'} />
        </WithRightMargin>
      )}
      {!!children && children}
    </button>
  );
};

// #region statics
BackToTopButton.defaultProps = {
  position: 'bottom-right',
};

BackToTopButton.displayName = 'BackToTopButton';
// #endregion

// #region helpers
function setPosition(position = 'bottom-right', refStyle = defaultStyle): any {
  const style = { ...refStyle };

  switch (position) {
    case 'bottom-right':
      style.right = sideOffset;
      style.left = '';
      return style;

    case 'bottom-left':
      style.right = '';
      style.left = sideOffset;
      return style;

    default:
      return refStyle;
  }
}
// #endregion

export default BackToTopButton;
