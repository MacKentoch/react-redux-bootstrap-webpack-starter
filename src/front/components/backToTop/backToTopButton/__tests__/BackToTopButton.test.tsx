import React from 'react';
import { shallow } from 'enzyme';
import BackToTopButton, { BackButtonPosition } from '../BackToTopButton';

describe('BackToTopButton component', () => {
  it('renders as expected', () => {
    const position: BackButtonPosition = 'bottom-left';
    const props = {
      position,
      onClick: () => {},
      motionStyle: {},
    };

    const component = shallow(
      <div>
        <BackToTopButton {...props}>
          <p>a child</p>
        </BackToTopButton>
      </div>,
    );

    expect(component).toMatchSnapshot();
  });
});
