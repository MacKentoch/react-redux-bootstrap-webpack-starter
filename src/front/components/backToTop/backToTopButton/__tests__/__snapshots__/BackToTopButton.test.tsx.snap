// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackToTopButton component renders as expected 1`] = `ShallowWrapper {}`;
